In [16]:
import pandas as pd
import numpy as np

In [17]:
df=pd.read_csv('word2vec_embedded.csv')

In [18]:
df.head()

label  word_count  num_stop_words  num_chars  num_punctuation_chars  \
0   ham          20               4        111                      9   
1   ham           6               0         29                      6   
2  spam          28               5        155                      5   
3   ham          11               2         49                      6   
4   ham          13               6         61                      2   

   word_count_after_preprocessing  num_chars_after_preprocessing  \
0                              16                             82   
1                               6                             23   
2                              20                            110   
3                               9                             35   
4                               8                             40   

                            word2vec_embedded_vector  
0  [-0.5803552, 0.0071118204, -0.9769345, -0.0825...  
1  [-0.42729643, -0.04335478, -0.79038095, 0.0076...  
2  [-0.21991925, 0.018497076, -0.65248924, 0.0868...  
3  [-0.62462306, 0.055353116, -1.4392701, -0.0687...  
4  [-0.33163285, -0.06703004, -1.2353566, -0.6881...

the embedded_vectors are stored as a list in a single column we have to store them in different column

In [19]:
df['word2vec_embedded_vector'][0]

'[-0.5803552, 0.0071118204, -0.9769345, -0.082504265, 0.14953394, -0.37411332, -0.1821864, 0.81884056, 0.10824226, -0.52496, -0.124138534, 0.14978546, -0.502542, -0.2748297, 0.024147969, 0.25342703, 0.1622787, -0.42769217, -0.060938314, -0.34580508, -0.15240194, 0.23503289, -0.19043575, -0.1769679, 0.07382312, -0.26729465, -0.7139932, -0.16784248, -0.59791154, -0.3681525, 0.4745909, -0.092171244, -0.15478797, 0.35238352, -0.71711016, -0.494439, -0.15903269, 0.4310765, -0.10301651, -0.060240526, 0.8406702, 0.103548154, -0.30228364, 0.06855834, -0.40137738, -0.5422219, -0.17131357, 0.31229904, 0.6384549, 0.60570717, 0.3803133, -0.42951235, 0.17598777, -0.30283654, 0.35648456, 0.251184, 0.300578, 0.026390169, -0.44820386, 0.23920389, 0.32895178, 0.27340835, -0.2559601, 0.24231653, -0.8286913, -0.20864369, -0.14372784, -0.06683013, 0.03796213, -0.02813201, -0.14570296, -0.54641014, 0.122905314, -0.43879467, 0.4644322, -0.2381447, 0.33804542, 0.18229583, -0.1034299, 0.19162747, 0.14924811, 

In [20]:
import ast
# Function to convert string to list
def convert_to_list(vector_str):
    return ast.literal_eval(vector_str)

In [21]:
# Apply the conversion to the 'word2vec_embedded_vector'
df['word2vec_embedded_vector'] = df['word2vec_embedded_vector'].apply(convert_to_list)

In [22]:
# Create a DataFrame from the list of vectors
expanded_df = df['word2vec_embedded_vector'].apply(pd.Series)

# Optionally, rename the columns
expanded_df.columns = [f'vector_dim_{i+1}' for i in range(expanded_df.shape[1])]

# Concatenate the new DataFrame with the original DataFrame (if needed)
final_df = pd.concat([df, expanded_df], axis=1)

# Display the final DataFrame
final_df.head()


label  word_count  num_stop_words  num_chars  num_punctuation_chars  \
0   ham          20               4        111                      9   
1   ham           6               0         29                      6   
2  spam          28               5        155                      5   
3   ham          11               2         49                      6   
4   ham          13               6         61                      2   

   word_count_after_preprocessing  num_chars_after_preprocessing  \
0                              16                             82   
1                               6                             23   
2                              20                            110   
3                               9                             35   
4                               8                             40   

                            word2vec_embedded_vector  vector_dim_1  \
0  [-0.5803552, 0.0071118204, -0.9769345, -0.0825...     -0.580355   
1  [-0.42729643, -0.04335478, -0.79038095, 0.0076...     -0.427296   
2  [-0.21991925, 0.018497076, -0.65248924, 0.0868...     -0.219919   
3  [-0.62462306, 0.055353116, -1.4392701, -0.0687...     -0.624623   
4  [-0.33163285, -0.06703004, -1.2353566, -0.6881...     -0.331633   

   vector_dim_2  ...  vector_dim_91  vector_dim_92  vector_dim_93  \
0      0.007112  ...       0.518878       0.497694      -0.090783   
1     -0.043355  ...       0.326726       0.445335      -0.144925   
2      0.018497  ...       0.153814       0.264300       0.085201   
3      0.055353  ...       0.667178       0.806222      -0.131574   
4     -0.067030  ...       0.939712       0.105435      -0.233439   

   vector_dim_94  vector_dim_95  vector_dim_96  vector_dim_97  vector_dim_98  \
0       0.009786       0.404360      -0.024591       0.145488      -0.293120   
1       0.105700       0.364651      -0.027538       0.004716      -0.113986   
2       0.226342       0.216582      -0.001449      -0.046417      -0.281185   
3       0.124313       0.451901      -0.234886      -0.067293      -0.029182   
4       0.112147      -0.122173       0.130926       0.006515      -0.164775   

   vector_dim_99  vector_dim_100  
0       0.501582       -0.175949  
1       0.182881       -0.253901  
2       0.236028       -0.549548  
3       0.448625       -0.219913  
4       1.332866       -0.211998  

[5 rows x 108 columns]

In [23]:
final_df.drop(columns='word2vec_embedded_vector',inplace=True)

In [24]:
final_df

label  word_count  num_stop_words  num_chars  num_punctuation_chars  \
0       ham          20               4        111                      9   
1       ham           6               0         29                      6   
2      spam          28               5        155                      5   
3       ham          11               2         49                      6   
4       ham          13               6         61                      2   
...     ...         ...             ...        ...                    ...   
14612   ham         131              59        745                     24   
14613   ham         121              62        670                     18   
14614   ham         141              72        770                      9   
14615   ham          86              47        448                     16   
14616   ham         260             139       1363                     35   

       word_count_after_preprocessing  num_chars_after_preprocessing  \
0                                  16                             82   
1                                   6                             23   
2                                  20                            110   
3                                   9                             35   
4                                   8                             40   
...                               ...                            ...   
14612                              74                            517   
14613                              63                            430   
14614                              69                            490   
14615                              41                            266   
14616                             126                            845   

       vector_dim_1  vector_dim_2  vector_dim_3  ...  vector_dim_91  \
0         -0.580355      0.007112     -0.976935  ...       0.518878   
1         -0.427296     -0.043355     -0.790381  ...       0.326726   
2         -0.219919      0.018497     -0.652489  ...       0.153814   
3         -0.624623      0.055353     -1.439270  ...       0.667178   
4         -0.331633     -0.067030     -1.235357  ...       0.939712   
...             ...           ...           ...  ...            ...   
14612     -0.342951     -0.075643     -1.186137  ...       0.801224   
14613     -0.297361     -0.135439     -1.199637  ...       0.920119   
14614     -0.336334     -0.135865     -1.247379  ...       0.946064   
14615     -0.438449     -0.073349     -1.116723  ...       0.790470   
14616     -0.414568     -0.125897     -1.152089  ...       0.947870   

       vector_dim_92  vector_dim_93  vector_dim_94  vector_dim_95  \
0           0.497694      -0.090783       0.009786       0.404360   
1           0.445335      -0.144925       0.105700       0.364651   
2           0.264300       0.085201       0.226342       0.216582   
3           0.806222      -0.131574       0.124313       0.451901   
4           0.105435      -0.233439       0.112147      -0.122173   
...              ...            ...            ...            ...   
14612       0.111723       0.127505       0.266898      -0.002344   
14613       0.085622       0.229993       0.302630      -0.120674   
14614       0.125366       0.388100       0.307724      -0.184426   
14615       0.178357       0.226579       0.250889      -0.079787   
14616       0.199644       0.153053       0.252963      -0.069350   

       vector_dim_96  vector_dim_97  vector_dim_98  vector_dim_99  \
0          -0.024591       0.145488      -0.293120       0.501582   
1          -0.027538       0.004716      -0.113986       0.182881   
2          -0.001449      -0.046417      -0.281185       0.236028   
3          -0.234886      -0.067293      -0.029182       0.448625   
4           0.130926       0.006515      -0.164775       1.332866   
...              ...            ...            ...            ...   
14612       0.037983      -0.183323      -0.352511 

In [25]:
final_df.to_csv('training_dataset.csv',index=False)

### Logistic Regression

In [26]:
from sklearn.preprocessing import LabelEncoder
label_encoder = LabelEncoder()
final_df['label'] = label_encoder.fit_transform(final_df['label'])
#0 for ham 1 for spam

In [27]:
final_df

label  word_count  num_stop_words  num_chars  num_punctuation_chars  \
0          0          20               4        111                      9   
1          0           6               0         29                      6   
2          1          28               5        155                      5   
3          0          11               2         49                      6   
4          0          13               6         61                      2   
...      ...         ...             ...        ...                    ...   
14612      0         131              59        745                     24   
14613      0         121              62        670                     18   
14614      0         141              72        770                      9   
14615      0          86              47        448                     16   
14616      0         260             139       1363                     35   

       word_count_after_preprocessing  num_chars_after_preprocessing  \
0                                  16                             82   
1                                   6                             23   
2                                  20                            110   
3                                   9                             35   
4                                   8                             40   
...                               ...                            ...   
14612                              74                            517   
14613                              63                            430   
14614                              69                            490   
14615                              41                            266   
14616                             126                            845   

       vector_dim_1  vector_dim_2  vector_dim_3  ...  vector_dim_91  \
0         -0.580355      0.007112     -0.976935  ...       0.518878   
1         -0.427296     -0.043355     -0.790381  ...       0.326726   
2         -0.219919      0.018497     -0.652489  ...       0.153814   
3         -0.624623      0.055353     -1.439270  ...       0.667178   
4         -0.331633     -0.067030     -1.235357  ...       0.939712   
...             ...           ...           ...  ...            ...   
14612     -0.342951     -0.075643     -1.186137  ...       0.801224   
14613     -0.297361     -0.135439     -1.199637  ...       0.920119   
14614     -0.336334     -0.135865     -1.247379  ...       0.946064   
14615     -0.438449     -0.073349     -1.116723  ...       0.790470   
14616     -0.414568     -0.125897     -1.152089  ...       0.947870   

       vector_dim_92  vector_dim_93  vector_dim_94  vector_dim_95  \
0           0.497694      -0.090783       0.009786       0.404360   
1           0.445335      -0.144925       0.105700       0.364651   
2           0.264300       0.085201       0.226342       0.216582   
3           0.806222      -0.131574       0.124313       0.451901   
4           0.105435      -0.233439       0.112147      -0.122173   
...              ...            ...            ...            ...   
14612       0.111723       0.127505       0.266898      -0.002344   
14613       0.085622       0.229993       0.302630      -0.120674   
14614       0.125366       0.388100       0.307724      -0.184426   
14615       0.178357       0.226579       0.250889      -0.079787   
14616       0.199644       0.153053       0.252963      -0.069350   

       vector_dim_96  vector_dim_97  vector_dim_98  vector_dim_99  \
0          -0.024591       0.145488      -0.293120       0.501582   
1          -0.027538       0.004716      -0.113986       0.182881   
2          -0.001449      -0.046417      -0.281185       0.236028   
3          -0.234886      -0.067293      -0.029182       0.448625   
4           0.130926       0.006515      -0.164775       1.332866   
...              ...            ...            ...            ...   
14612       0.037983      -0.183323     

In [28]:
x=final_df.drop(columns='label')

In [29]:
x

word_count  num_stop_words  num_chars  num_punctuation_chars  \
0              20               4        111                      9   
1               6               0         29                      6   
2              28               5        155                      5   
3              11               2         49                      6   
4              13               6         61                      2   
...           ...             ...        ...                    ...   
14612         131              59        745                     24   
14613         121              62        670                     18   
14614         141              72        770                      9   
14615          86              47        448                     16   
14616         260             139       1363                     35   

       word_count_after_preprocessing  num_chars_after_preprocessing  \
0                                  16                             82   
1                                   6                             23   
2                                  20                            110   
3                                   9                             35   
4                                   8                             40   
...                               ...                            ...   
14612                              74                            517   
14613                              63                            430   
14614                              69                            490   
14615                              41                            266   
14616                             126                            845   

       vector_dim_1  vector_dim_2  vector_dim_3  vector_dim_4  ...  \
0         -0.580355      0.007112     -0.976935     -0.082504  ...   
1         -0.427296     -0.043355     -0.790381      0.007617  ...   
2         -0.219919      0.018497     -0.652489      0.086807  ...   
3         -0.624623      0.055353     -1.439270     -0.068720  ...   
4         -0.331633     -0.067030     -1.235357     -0.688138  ...   
...             ...           ...           ...           ...  ...   
14612     -0.342951     -0.075643     -1.186137     -0.375801  ...   
14613     -0.297361     -0.135439     -1.199637     -0.433341  ...   
14614     -0.336334     -0.135865     -1.247379     -0.403766  ...   
14615     -0.438449     -0.073349     -1.116723     -0.389665  ...   
14616     -0.414568     -0.125897     -1.152089     -0.421651  ...   

       vector_dim_91  vector_dim_92  vector_dim_93  vector_dim_94  \
0           0.518878       0.497694      -0.090783       0.009786   
1           0.326726       0.445335      -0.144925       0.105700   
2           0.153814       0.264300       0.085201       0.226342   
3           0.667178       0.806222      -0.131574       0.124313   
4           0.939712       0.105435      -0.233439       0.112147   
...              ...            ...            ...            ...   
14612       0.801224       0.111723       0.127505       0.266898   
14613       0.920119       0.085622       0.229993       0.302630   
14614       0.946064       0.125366       0.388100       0.307724   
14615       0.790470       0.178357       0.226579       0.250889   
14616       0.947870       0.199644       0.153053       0.252963   

       vector_dim_95  vector_dim_96  vector_dim_97  vector_dim_98  \
0           0.404360      -0.024591       0.145488      -0.293120   
1           0.364651      -0.027538       0.004716      -0.113986   
2           0.216582      -0.001449      -0.046417      -0.281185   
3           0.451901      -0.234886      -0.067293      -0.029182   
4          -0.122173       0.130926       0.006515      -0.164775   
...              ...            ...            ...            ...   
14612      -0.002344       0.037983      -0.183323      -0.352511   
14613      -0.120674       0.033732      -0.367308      -0.270141  

In [30]:
y=final_df['label']

In [31]:
y

0        0
1        0
2        1
3        0
4        0
        ..
14612    0
14613    0
14614    0
14615    0
14616    0
Name: label, Length: 14617, dtype: int64

Logistic Regression

In [32]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
from sklearn.preprocessing import StandardScaler

In [33]:
# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(x,y, test_size=0.2, random_state=42)


In [34]:
# Initialize the StandardScaler
scaler = StandardScaler()

# Fit the scaler on the training data and transform both training and testing data
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)


In [35]:
# Initialize the Logistic Regression model
LR = LogisticRegression(max_iter=500)

# Train the model
LR.fit(X_train_scaled, y_train)


LogisticRegression(max_iter=500)

In [36]:
# Make predictions on the scaled test set
y_pred = LR.predict(X_test_scaled)

In [37]:
# Calculate accuracy
accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy: {accuracy:.2f}")

Accuracy: 0.89


In [38]:
# Print confusion matrix and classification report
print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred))


Confusion Matrix:
[[2142  113]
 [ 217  452]]


In [39]:
print("\nClassification Report:")
print(classification_report(y_test, y_pred))


Classification Report:
              precision    recall  f1-score   support

           0       0.91      0.95      0.93      2255
           1       0.80      0.68      0.73       669

    accuracy                           0.89      2924
   macro avg       0.85      0.81      0.83      2924
weighted avg       0.88      0.89      0.88      2924



In [40]:
from sklearn.metrics import roc_auc_score
y_pred_proba = LR.predict_proba(X_test)
roc_auc_score(y_test, y_pred_proba[:, 0])

e:\Programming & DS\# Projects\Phishing Email Detection\myenv\Lib\site-packages\sklearn\base.py:486: UserWarning: X has feature names, but LogisticRegression was fitted without feature names
  warnings.warn(


np.float64(0.6191350229849628)

In [41]:
roc_auc_score(y_test, y_pred_proba[:, 1])

np.float64(0.40208206974038757)

In [43]:
import statsmodels.api as sm

X = sm.add_constant(X_train)  # Add a constant term (intercept)
model = sm.Logit(y_train, X).fit()  
print(model.summary())  


Optimization terminated successfully.
         Current function value: 0.240779
         Iterations 11
                           Logit Regression Results                           
Dep. Variable:                  label   No. Observations:                11693
Model:                          Logit   Df Residuals:                    11586
Method:                           MLE   Df Model:                          106
Date:                Mon, 11 Nov 2024   Pseudo R-squ.:                  0.5510
Time:                        12:44:27   Log-Likelihood:                -2815.4
converged:                       True   LL-Null:                       -6270.0
Covariance Type:            nonrobust   LLR p-value:                     0.000
                                     coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------------------------
const                             -0.6666      0.164     -4.075    



### 1. **Model Information:**
   - **Dependent Variable (`Dep. Variable`)**: `label`, which refers to the target or outcome variable you're predicting.
   - **Number of Observations (`No. Observations`)**: 11,693, indicating the number of samples used to train the model.
   - **Model (`Model`)**: The logistic regression model, which uses Maximum Likelihood Estimation (`MLE`).
   - **Residual Degrees of Freedom (`Df Residuals`)**: 11,586, which is the difference between the number of observations and the number of model parameters (106).
   - **Pseudo R-squared (`Pseudo R-squ.`)**: 0.5550, a goodness-of-fit measure. This suggests that around **55.50%** of the variance in the target variable is explained by the model.

### 2. **Log-Likelihood Values:**
   - **Current function value (Log-Likelihood)**: -2,789.9, which is the final value of the log-likelihood after optimization. Lower values indicate a better fit of the model.
   - **LL-Null**: -6,270.0, the log-likelihood for a model with no predictors (only the intercept).
   - **LLR p-value**: 0.000, which indicates that the model as a whole is statistically significant compared to a null model.

### 3. **Coefficients and Interpretation:**
   For each feature, the model provides:
   - **Coefficient (`coef`)**: The impact of that feature on the log-odds of the outcome. A positive coefficient increases the odds of the event happening, while a negative coefficient decreases it.
   - **Standard Error (`std err`)**: The variability of the coefficient estimate.
   - **Z-value (`z`)**: The test statistic for testing whether the coefficient is different from 0.
   - **P>|z|**: The p-value, showing the statistical significance of the coefficient. Typically, a p-value below 0.05 indicates that the feature is statistically significant.
   - **Confidence Interval**: The 95% confidence interval for the coefficient, which gives a range within which the true value of the coefficient likely lies.

   Let's interpret some specific features:
   - **`const`**: The intercept has a significant negative value (-0.6502) with a p-value of 0.000, meaning it’s statistically significant.
   - **`word_count`**: The coefficient (0.0007) is positive, but the p-value (0.774) is large, indicating that this feature is not statistically significant.
   - **`num_chars`**: The coefficient (-0.0010) suggests a negative impact, but with a p-value of 0.350, it's not statistically significant.
   - **`vector_dim_1`**: The coefficient (-4.1892) is large in magnitude but has a high p-value (0.325), meaning it’s also not statistically significant.

   Most of the features shown have **high p-values** (above 0.05), indicating they are **not statistically significant** predictors in this model.

### 4. **Quasi-Separation Warning:**
   - The message **"Possibly complete quasi-separation"** suggests that a fraction (14%) of observations are perfectly predicted by the model. This can lead to **complete quasi-separation**, meaning that some features can perfectly predict the outcome for certain samples.
   - **Implications**: This might cause instability in the estimated coefficients, especially with large positive or negative values. It also suggests that some parameters may not be identified, meaning their effects can’t be accurately estimated.

### 5. **Overall Model Fit and Significance:**
   - The model as a whole is significant, as indicated by the LLR p-value of 0.000.
   - However, most individual features do not appear to be significant, suggesting that they may not contribute much predictive power individually.
   - The Pseudo R-squared of 0.555 indicates a decent fit, but there's room for improvement in explaining the variability in the outcome.

### Next Steps:
   - You may want to assess whether some features are collinear or unnecessary and try simplifying the model by removing insignificant predictors.
   - Address the **quasi-separation issue** by possibly removing or regularizing the features causing this, or using penalized logistic regression (e.g., Lasso or Ridge).



### Random Forest

In [44]:
x

word_count  num_stop_words  num_chars  num_punctuation_chars  \
0              20               4        111                      9   
1               6               0         29                      6   
2              28               5        155                      5   
3              11               2         49                      6   
4              13               6         61                      2   
...           ...             ...        ...                    ...   
14612         131              59        745                     24   
14613         121              62        670                     18   
14614         141              72        770                      9   
14615          86              47        448                     16   
14616         260             139       1363                     35   

       word_count_after_preprocessing  num_chars_after_preprocessing  \
0                                  16                             82   
1                                   6                             23   
2                                  20                            110   
3                                   9                             35   
4                                   8                             40   
...                               ...                            ...   
14612                              74                            517   
14613                              63                            430   
14614                              69                            490   
14615                              41                            266   
14616                             126                            845   

       vector_dim_1  vector_dim_2  vector_dim_3  vector_dim_4  ...  \
0         -0.580355      0.007112     -0.976935     -0.082504  ...   
1         -0.427296     -0.043355     -0.790381      0.007617  ...   
2         -0.219919      0.018497     -0.652489      0.086807  ...   
3         -0.624623      0.055353     -1.439270     -0.068720  ...   
4         -0.331633     -0.067030     -1.235357     -0.688138  ...   
...             ...           ...           ...           ...  ...   
14612     -0.342951     -0.075643     -1.186137     -0.375801  ...   
14613     -0.297361     -0.135439     -1.199637     -0.433341  ...   
14614     -0.336334     -0.135865     -1.247379     -0.403766  ...   
14615     -0.438449     -0.073349     -1.116723     -0.389665  ...   
14616     -0.414568     -0.125897     -1.152089     -0.421651  ...   

       vector_dim_91  vector_dim_92  vector_dim_93  vector_dim_94  \
0           0.518878       0.497694      -0.090783       0.009786   
1           0.326726       0.445335      -0.144925       0.105700   
2           0.153814       0.264300       0.085201       0.226342   
3           0.667178       0.806222      -0.131574       0.124313   
4           0.939712       0.105435      -0.233439       0.112147   
...              ...            ...            ...            ...   
14612       0.801224       0.111723       0.127505       0.266898   
14613       0.920119       0.085622       0.229993       0.302630   
14614       0.946064       0.125366       0.388100       0.307724   
14615       0.790470       0.178357       0.226579       0.250889   
14616       0.947870       0.199644       0.153053       0.252963   

       vector_dim_95  vector_dim_96  vector_dim_97  vector_dim_98  \
0           0.404360      -0.024591       0.145488      -0.293120   
1           0.364651      -0.027538       0.004716      -0.113986   
2           0.216582      -0.001449      -0.046417      -0.281185   
3           0.451901      -0.234886      -0.067293      -0.029182   
4          -0.122173       0.130926       0.006515      -0.164775   
...              ...            ...            ...            ...   
14612      -0.002344       0.037983      -0.183323      -0.352511   
14613      -0.120674       0.033732      -0.367308      -0.270141  

In [45]:
y

0        0
1        0
2        1
3        0
4        0
        ..
14612    0
14613    0
14614    0
14615    0
14616    0
Name: label, Length: 14617, dtype: int64

## Random Forest Classifier

In [46]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

# Splitting the data
X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42)

# Initializing the Random Forest model with class_weight='balanced'
rf_model = RandomForestClassifier(n_estimators=100,class_weight='balanced', random_state=42)

# Fitting the model
rf_model.fit(X_train, y_train)

# Predictions
y_pred = rf_model.predict(X_test)

# Evaluating the model
print(classification_report(y_test, y_pred))


              precision    recall  f1-score   support

           0       0.95      0.98      0.97      2255
           1       0.94      0.83      0.88       669

    accuracy                           0.95      2924
   macro avg       0.95      0.91      0.93      2924
weighted avg       0.95      0.95      0.95      2924



In [47]:
# Print confusion matrix and classification report
print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred))


Confusion Matrix:
[[2221   34]
 [ 114  555]]


In [48]:
y_pred_proba = rf_model.predict_proba(X_test)
roc_auc_score(y_test, y_pred_proba[:, 0])

np.float64(0.016031472993082968)

In [49]:
roc_auc_score(y_test, y_pred_proba[:, 1])

np.float64(0.983968527006917)

### Decision Tree Classifier

In [50]:
X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42)

In [51]:
from sklearn.tree import DecisionTreeClassifier
dt_classifier = DecisionTreeClassifier(random_state=42)
# Train the model
dt_classifier.fit(X_train, y_train)

DecisionTreeClassifier(random_state=42)

In [52]:
# Make predictions
y_pred = dt_classifier.predict(X_test)

In [53]:
print(confusion_matrix(y_test, y_pred))

[[2126  129]
 [ 140  529]]


In [54]:
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       0.94      0.94      0.94      2255
           1       0.80      0.79      0.80       669

    accuracy                           0.91      2924
   macro avg       0.87      0.87      0.87      2924
weighted avg       0.91      0.91      0.91      2924



In [55]:
y_pred_proba = dt_classifier.predict_proba(X_test)
roc_auc_score(y_test, y_pred_proba[:, 0])

np.float64(0.13323688597668693)

In [56]:
roc_auc_score(y_test, y_pred_proba[:, 1])

np.float64(0.8667631140233131)

### XGBOOST

In [57]:
from sklearn.model_selection import train_test_split

# Example: X is your features and y is your target
X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42)


In [59]:
from xgboost import XGBClassifier

# Calculate scale_pos_weight (ratio of majority class to minority class)
ratio = sum(y_train == 0) / sum(y_train == 1)

# Create the XGBoost model
XG_Boost_model = XGBClassifier(
    scale_pos_weight=ratio,  # Handle class imbalance
    n_estimators=100,        # Number of trees
    max_depth=4,             # Maximum depth of the trees
    learning_rate=0.1,       # Step size shrinkage
    objective='binary:logistic',  # Binary classification
    random_state=42,         # Random state for reproducibility
    eval_metric='logloss'    # Evaluation metric for binary classification
)


In [60]:
ratio

3.3958646616541355

In [61]:
# Train the model
XG_Boost_model.fit(X_train, y_train)


XGBClassifier(base_score=None, booster=None, callbacks=None,
              colsample_bylevel=None, colsample_bynode=None,
              colsample_bytree=None, device=None, early_stopping_rounds=None,
              enable_categorical=False, eval_metric='logloss',
              feature_types=None, gamma=None, grow_policy=None,
              importance_type=None, interaction_constraints=None,
              learning_rate=0.1, max_bin=None, max_cat_threshold=None,
              max_cat_to_onehot=None, max_delta_step=None, max_depth=4,
              max_leaves=None, min_child_weight=None, missing=nan,
              monotone_constraints=None, multi_strategy=None, n_estimators=100,
              n_jobs=None, num_parallel_tree=None, random_state=42, ...)

In [62]:
# Predicting on the test set
y_pred = XG_Boost_model.predict(X_test)


In [63]:
from sklearn.metrics import classification_report, confusion_matrix, roc_auc_score

# Classification report
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.98      0.94      0.96      2255
           1       0.82      0.94      0.88       669

    accuracy                           0.94      2924
   macro avg       0.90      0.94      0.92      2924
weighted avg       0.95      0.94      0.94      2924



In [64]:
# Confusion matrix
print(confusion_matrix(y_test, y_pred))

[[2120  135]
 [  40  629]]


True Positives (TP) = 625:
These are cases where the model correctly predicted email as spam (positive class).

True Negatives (TN) = 2130:
These are cases where the model correctly predicted email as ham (negative class).

False Positives (FP) = 125:
These are cases where the model incorrectly predicted email as spam (positive class) when the actual email was ham.

False Negatives (FN) = 44:
These are cases where the model incorrectly predicted email as ham (negative class) when the actual email was spam.

In [65]:
# AUC-ROC score
y_pred_proba = XG_Boost_model.predict_proba(X_test)[:, 0]
print(f"AUC-ROC: {roc_auc_score(y_test, y_pred_proba)}")

AUC-ROC: 0.016891213347518717


In [66]:
# AUC-ROC score
y_pred_proba = XG_Boost_model.predict_proba(X_test)[:, 1]
print(f"AUC-ROC: {roc_auc_score(y_test, y_pred_proba)}")

AUC-ROC: 0.9831087866524812


## Stacking

In [67]:
from sklearn.ensemble import StackingClassifier
from sklearn.neighbors import KNeighborsClassifier

In [68]:
# Meta-learner
knn_meta_learner = KNeighborsClassifier(n_neighbors=5)

In [69]:
# Create the StackingClassifier with RandomForestClassifier and XGBoostClassifier as base models, and KNN as meta-learner
stacking_model = StackingClassifier(
    estimators=[
        ('RandomForestClassifier', rf_model),
        ('XGBoostClassifier', XG_Boost_model)
    ],
    final_estimator=knn_meta_learner,
    cv=5
)

# Train the stacking model
stacking_model.fit(X_train, y_train)

# Make predictions on the test data
y_pred = stacking_model.predict(X_test)

# Generate classification report
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.96      0.97      0.96      2255
           1       0.90      0.86      0.88       669

    accuracy                           0.94      2924
   macro avg       0.93      0.91      0.92      2924
weighted avg       0.94      0.94      0.94      2924



In [70]:
# Confusion matrix
confusion_matrix(y_test, y_pred)

array([[2189,   66],
       [  97,  572]])

In [71]:
# AUC-ROC score
y_pred_proba = stacking_model.predict_proba(X_test)[:, 0]
print(f"AUC-ROC: {roc_auc_score(y_test, y_pred_proba)}")

AUC-ROC: 0.034629572549292546


In [72]:
# AUC-ROC score
y_pred_proba = stacking_model.predict_proba(X_test)[:, 1]
print(f"AUC-ROC: {roc_auc_score(y_test, y_pred_proba)}")

AUC-ROC: 0.9653704274507076


In [73]:
import pandas as pd
from sklearn.metrics import classification_report, confusion_matrix

# Assume you have your trained models in a dictionary
models = {
    'Logistic Regression': LR,
    'Random Forest': rf_model,
    'XGBClassifier': XG_Boost_model,
    'Decision Tree Classifier': dt_classifier,
    'Stacking Model': stacking_model
}

# Initialize lists to store the results
results = []

# Loop through each model to evaluate
for model_name, model in models.items():
    # Make predictions
    y_pred = model.predict(X_test)  # Ensure X_test is defined
    y_true = y_test  # Ensure y_test is defined

    # Get classification report
    report = classification_report(y_true, y_pred, output_dict=True)

    # Get confusion matrix
    cm = confusion_matrix(y_true, y_pred)

    # Extract relevant metrics safely
    accuracy = report['accuracy']
    precision = report['1']['precision'] if '1' in report else 0.0  # Handle missing class
    recall = report['1']['recall'] if '1' in report else 0.0  # Handle missing class
    f1_score = report['1']['f1-score'] if '1' in report else 0.0  # Handle missing class
    type_1_error = cm[0][1]  # False Positive
    type_2_error = cm[1][0]  # False Negative

    # Append the results
    results.append({
        'Model': model_name,
        'Accuracy': accuracy,
        'Precision': precision,
        'Recall': recall,
        'F1 Score': f1_score,
        'Type 1 Error (FP)': type_1_error,
        'Type 2 Error (FN)': type_2_error,
        'Confusion Matrix': cm.tolist()  # Convert to list for better readability
    })

# Create a DataFrame from the results
results_df = pd.DataFrame(results)



e:\Programming & DS\# Projects\Phishing Email Detection\myenv\Lib\site-packages\sklearn\base.py:486: UserWarning: X has feature names, but LogisticRegression was fitted without feature names
  warnings.warn(
e:\Programming & DS\# Projects\Phishing Email Detection\myenv\Lib\site-packages\sklearn\metrics\_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
e:\Programming & DS\# Projects\Phishing Email Detection\myenv\Lib\site-packages\sklearn\metrics\_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
e:\Programming & DS\# Projects\Phishing Email Detection\myenv\Lib\site-package

In [74]:
results_df

Model  Accuracy  Precision    Recall  F1 Score  \
0       Logistic Regression  0.771204   0.000000  0.000000  0.000000   
1             Random Forest  0.949384   0.942275  0.829596  0.882353   
2             XGBClassifier  0.940150   0.823298  0.940209  0.877879   
3  Decision Tree Classifier  0.908003   0.803951  0.790732  0.797287   
4            Stacking Model  0.944254   0.896552  0.855007  0.875287   

   Type 1 Error (FP)  Type 2 Error (FN)           Confusion Matrix  
0                  0                669      [[2255, 0], [669, 0]]  
1                 34                114   [[2221, 34], [114, 555]]  
2                135                 40   [[2120, 135], [40, 629]]  
3                129                140  [[2126, 129], [140, 529]]  
4                 66                 97    [[2189, 66], [97, 572]]



### Analysis and Recommendations

1. **Overall Performance**:
   - **Random Forest** continues to perform the best overall with the highest precision and F1-scores for both classes.
   - **XGBoost** follows closely, demonstrating strong performance, especially in precision for not spam (class 0).
   - **Decision Tree** shows reasonable performance but lags behind the other models, particularly in precision for spam (class 1).
   - **Logistic Regression** has the lowest metrics overall, particularly in recall for spam (class 1), indicating it misses many spam messages.

2. **Precision**:
   - **XGBoost** exhibits the highest precision for not spam (0.98) and a competitive score for spam (0.83).
   - **Random Forest** has a very high precision for spam (0.94) but lower for not spam compared to XGBoost.
   - **Decision Tree** has a good precision for not spam (0.94) but a lower precision for spam (0.80).

3. **Recall**:
   - **Random Forest** maintains the highest recall for not spam (0.98) and a decent score for spam (0.84).
   - **Decision Tree** has good recall for not spam (0.94) but lower recall for spam (0.79).
   - **Logistic Regression** has the lowest recall for spam (0.69), which indicates it misses many spam emails.

4. **F1-Score**:
   - The **Random Forest** achieves the highest F1-scores overall, indicating a good balance between precision and recall for both classes.
   - **XGBoost** also has high F1-scores, particularly for not spam (0.96).
   - **Decision Tree** has a good F1-score for not spam (0.94) but lower for spam (0.80), while **Logistic Regression** has the lowest F1-scores, especially for spam (0.75).

### Conclusion

Based on the performance metrics:

- **Random Forest** is the best model overall for this email spam classification task, showing strong performance in precision, recall, and F1-scores across both classes.
- **XGBoost** is a close second, especially notable for its high precision for not spam.
- **Decision Tree** performs reasonably well but does not match the other models in terms of spam detection capabilities.
- **Logistic Regression** is the weakest performer and may not be suitable for applications where spam detection is critical.



In [75]:
import pickle

# Save the model
with open('random_forest__baseline_model.pkl', 'wb') as file:
    pickle.dump(rf_model, file)